File: ascii_city/infrastructure/tile_codec.py
"""Binary tile container. See ``docs/world-format.md`` for the byte layout.

One 128 x 128 tile encodes to roughly 55 KB and compresses to well under
10 KB, because the three cell layers are highly repetitive. The client decodes
this inside a Web Worker so the game loop never stalls on it.
"""

from __future__ import annotations

import struct
from typing import Sequence

from ..domain.errors import WorldDataError
from ..domain.world import Building, Prop, Road, SpawnPoint, WorldTile
from .quantise import round_half_up

MAGIC = b"ACT1"
FORMAT_VERSION = 1

_HEADER = struct.Struct("<4sHHiiHfI")
_BUILDING_TAIL = struct.Struct("<BBBBBBBBB")
_PROP = struct.Struct("<HHHB")
_SPAWN = struct.Struct("<HHH")

_ANGLE_UNITS = 65536.0
_TAU = 6.283185307179586


class _Reader:
    __slots__ = ("_data", "_offset")

    def __init__(self, data: bytes) -> None:
        self._data = data
        self._offset = 0

    def take(self, size: int) -> bytes:
        end = self._offset + size
        if end > len(self._data):
            raise WorldDataError("Tile payload is truncated.")
        chunk = self._data[self._offset : end]
        self._offset = end
        return chunk

    def unpack(self, layout: struct.Struct) -> tuple:
        return layout.unpack(self.take(layout.size))

    def u8(self) -> int:
        return self.take(1)[0]

    def u16(self) -> int:
        return int.from_bytes(self.take(2), "little")

    def i16(self) -> int:
        return int.from_bytes(self.take(2), "little", signed=True)

    def text(self, length: int) -> str:
        return self.take(length).decode("utf-8")


def encode_tile(tile: WorldTile) -> bytes:
    out = bytearray()
    identifier = tile.id.encode("utf-8")
    out += _HEADER.pack(
        MAGIC,
        FORMAT_VERSION,
        0,
        tile.tile_x,
        tile.tile_y,
        tile.cells,
        tile.cell_size,
        tile.version,
    )
    out += len(identifier).to_bytes(2, "little")
    out += identifier

    out += tile.collision
    out += tile.heights
    out += tile.styles

    out += len(tile.buildings).to_bytes(2, "little")
    for building in tile.buildings:
        out += building.id.to_bytes(2, "little")
        out.append(building.vertex_count)
        for value in building.footprint:
            out += int(value).to_bytes(2, "little", signed=True)
        out += _BUILDING_TAIL.pack(
            min(255, building.height),
            min(255, building.min_height),
            min(255, building.levels),
            building.roof_type,
            building.category,
            building.facade_style,
            building.window_style,
            building.color,
            (1 if building.walkable else 0) | (2 if building.interior_id else 0),
        )

    out += len(tile.roads).to_bytes(2, "little")
    for road in tile.roads:
        name = (road.name or "").encode("utf-8")[:255]
        out += road.id.to_bytes(2, "little")
        out.append(road.type)
        out.append(min(255, max(1, int(round(road.width * 10)))))
        out.append(road.surface_style)
        out.append(len(road.centerline) // 2)
        for value in road.centerline:
            out += int(value).to_bytes(2, "little", signed=True)
        out.append(len(name))
        out += name

    out += len(tile.props).to_bytes(2, "little")
    for prop in tile.props:
        out += _PROP.pack(prop.id, prop.x, prop.y, prop.kind)

    out += len(tile.spawn_points).to_bytes(2, "little")
    for spawn in tile.spawn_points:
        out += _SPAWN.pack(spawn.x, spawn.y, _encode_heading(spawn.heading))

    return bytes(out)


def decode_tile(payload: bytes) -> WorldTile:
    reader = _Reader(payload)
    magic, version, _flags, tile_x, tile_y, cells, cell_size, world_version = reader.unpack(_HEADER)
    if magic != MAGIC:
        raise WorldDataError("Not an ASCII City tile payload.")
    if version != FORMAT_VERSION:
        raise WorldDataError(f"Unsupported tile format version {version}.")
    identifier = reader.text(reader.u16())

    area = cells * cells
    collision = bytearray(reader.take(area))
    heights = bytearray(reader.take(area))
    styles = bytearray(reader.take(area))

    buildings: list[Building] = []
    for _ in range(reader.u16()):
        building_id = reader.u16()
        vertex_count = reader.u8()
        footprint = tuple(reader.i16() for _ in range(vertex_count * 2))
        (
            height,
            min_height,
            levels,
            roof_type,
            category,
            facade_style,
            window_style,
            color,
            flags,
        ) = reader.unpack(_BUILDING_TAIL)
        buildings.append(
            Building(
                id=building_id,
                footprint=footprint,
                height=height,
                min_height=min_height,
                levels=levels,
                roof_type=roof_type,
                category=category,
                facade_style=facade_style,
                window_style=window_style,
                color=color,
                walkable=bool(flags & 1),
                interior_id="pending" if flags & 2 else None,
            )
        )

    roads: list[Road] = []
    for _ in range(reader.u16()):
        road_id = reader.u16()
        road_type = reader.u8()
        width = reader.u8() / 10.0
        surface_style = reader.u8()
        point_count = reader.u8()
        centerline = tuple(reader.i16() for _ in range(point_count * 2))
        name = reader.text(reader.u8()) or None
        roads.append(
            Road(
                id=road_id,
                centerline=centerline,
                width=width,
                type=road_type,
                walkable=True,
                surface_style=surface_style,
                name=name,
            )
        )

    props = tuple(
        Prop(*reader.unpack(_PROP)) for _ in range(reader.u16())
    )

    spawns: list[SpawnPoint] = []
    for _ in range(reader.u16()):
        x, y, heading = reader.unpack(_SPAWN)
        spawns.append(SpawnPoint(x=x, y=y, heading=heading / _ANGLE_UNITS * _TAU))

    return WorldTile(
        id=identifier,
        version=world_version,
        tile_x=tile_x,
        tile_y=tile_y,
        cells=cells,
        cell_size=cell_size,
        collision=collision,
        heights=heights,
        styles=styles,
        buildings=tuple(buildings),
        roads=tuple(roads),
        props=props,
        spawn_points=tuple(spawns),
    )


def _encode_heading(heading: float) -> int:
    return round_half_up((heading % _TAU) / _TAU * _ANGLE_UNITS) & 0xFFFF


def describe_tiles(tiles: Sequence[WorldTile]) -> dict[str, object]:
    """Small JSON summary used by the world metadata endpoint and the tests."""
    return {
        "count": len(tiles),
        "buildings": sum(len(tile.buildings) for tile in tiles),
        "roads": sum(len(tile.roads) for tile in tiles),
        "props": sum(len(tile.props) for tile in tiles),
        "spawnPoints": sum(len(tile.spawn_points) for tile in tiles),
    }

File: ascii_city/frontend/src/world/tileCodec.ts
/**
 * Decoder for the binary tile container produced by
 * `ascii_city/infrastructure/tile_codec.py`. `docs/world-format.md` is the
 * normative byte layout; this file must not diverge from it.
 *
 * Decoding runs inside a Web Worker, so it returns plain typed arrays that
 * transfer to the main thread without a copy.
 */

import type { Building, Prop, Road, SpawnPoint, WorldTile } from '../domain/types'
import { TAU } from '../domain/constants'

export const MAGIC = 0x31544341 // "ACT1" read as a little-endian u32
export const FORMAT_VERSION = 1

const ANGLE_UNITS = 65536

export class TileFormatError extends Error {}

class Reader {
  private offset = 0
  private readonly view: DataView
  private readonly text = new TextDecoder()

  constructor(private readonly bytes: Uint8Array) {
    this.view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
  }

  private require(size: number): number {
    const at = this.offset
    if (at + size > this.bytes.byteLength) {
      throw new TileFormatError('Tile payload is truncated.')
    }
    this.offset = at + size
    return at
  }

  u8(): number {
    return this.view.getUint8(this.require(1))
  }

  u16(): number {
    return this.view.getUint16(this.require(2), true)
  }

  i16(): number {
    return this.view.getInt16(this.require(2), true)
  }

  i32(): number {
    return this.view.getInt32(this.require(4), true)
  }

  u32(): number {
    return this.view.getUint32(this.require(4), true)
  }

  f32(): number {
    return this.view.getFloat32(this.require(4), true)
  }

  /** A copy, because the caller keeps it after the source buffer is released. */
  layer(size: number): Uint8Array {
    const at = this.require(size)
    return this.bytes.slice(at, at + size)
  }

  string(length: number): string {
    const at = this.require(length)
    return this.text.decode(this.bytes.subarray(at, at + length))
  }

  i16Array(count: number): Int16Array {
    const values = new Int16Array(count)
    for (let index = 0; index < count; index += 1) values[index] = this.i16()
    return values
  }
}

export function decodeTile(payload: ArrayBuffer | Uint8Array): WorldTile {
  const bytes = payload instanceof Uint8Array ? payload : new Uint8Array(payload)
  const reader = new Reader(bytes)

  if (reader.u32() !== MAGIC) throw new TileFormatError('Not an ASCII City tile payload.')
  const version = reader.u16()
  if (version !== FORMAT_VERSION) {
    throw new TileFormatError(`Unsupported tile format version ${version}.`)
  }
  reader.u16() // flags, reserved
  const tileX = reader.i32()
  const tileY = reader.i32()
  const cells = reader.u16()
  const cellSize = reader.f32()
  const worldVersion = reader.u32()
  const id = reader.string(reader.u16())

  const area = cells * cells
  const collision = reader.layer(area)
  const heights = reader.layer(area)
  const styles = reader.layer(area)

  const buildings: Building[] = []
  for (let index = reader.u16(); index > 0; index -= 1) {
    const buildingId = reader.u16()
    const vertexCount = reader.u8()
    const footprint = reader.i16Array(vertexCount * 2)
    const height = reader.u8()
    const minHeight = reader.u8()
    const levels = reader.u8()
    const roofType = reader.u8()
    const category = reader.u8()
    const facadeStyle = reader.u8()
    const windowStyle = reader.u8()
    const color = reader.u8()
    const flags = reader.u8()
    buildings.push({
      id: buildingId,
      footprint,
      height,
      minHeight,
      levels,
      roofType,
      category,
      facadeStyle,
      windowStyle,
      color,
      walkable: (flags & 1) !== 0,
      hasInterior: (flags & 2) !== 0,
    })
  }

  const roads: Road[] = []
  for (let index = reader.u16(); index > 0; index -= 1) {
    const roadId = reader.u16()
    const type = reader.u8()
    const width = reader.u8() / 10
    const surfaceStyle = reader.u8()
    const centerline = reader.i16Array(reader.u8() * 2)
    const nameLength = reader.u8()
    const name = nameLength > 0 ? reader.string(nameLength) : null
    roads.push({ id: roadId, centerline, width, type, surfaceStyle, name })
  }

  const props: Prop[] = []
  for (let index = reader.u16(); index > 0; index -= 1) {
    props.push({ id: reader.u16(), x: reader.u16(), y: reader.u16(), kind: reader.u8() })
  }

  const spawnPoints: SpawnPoint[] = []
  for (let index = reader.u16(); index > 0; index -= 1) {
    spawnPoints.push({
      x: reader.u16(),
      y: reader.u16(),
      heading: (reader.u16() / ANGLE_UNITS) * TAU,
    })
  }

  return {
    id,
    version: worldVersion,
    tileX,
    tileY,
    cells,
    cellSize,
    collision,
    heights,
    styles,
    buildings,
    roads,
    props,
    spawnPoints,
  }
}
